File: clients/Cpp/model/BuildAction.cpp
#include "BuildAction.hpp"

BuildAction::BuildAction() { }
BuildAction::BuildAction(EntityType entityType, Vec2Int position) : entityType(entityType), position(position) { }
BuildAction BuildAction::readFrom(InputStream& stream) {
    BuildAction result;
    switch (stream.readInt()) {
    case 0:
        result.entityType = EntityType::WALL;
        break;
    case 1:
        result.entityType = EntityType::HOUSE;
        break;
    case 2:
        result.entityType = EntityType::BUILDER_BASE;
        break;
    case 3:
        result.entityType = EntityType::BUILDER_UNIT;
        break;
    case 4:
        result.entityType = EntityType::MELEE_BASE;
        break;
    case 5:
        result.entityType = EntityType::MELEE_UNIT;
        break;
    case 6:
        result.entityType = EntityType::RANGED_BASE;
        break;
    case 7:
        result.entityType = EntityType::RANGED_UNIT;
        break;
    case 8:
        result.entityType = EntityType::RESOURCE;
        break;
    case 9:
        result.entityType = EntityType::TURRET;
        break;
    default:
        throw std::runtime_error("Unexpected tag value");
    }
    result.position = Vec2Int::readFrom(stream);
    return result;
}
void BuildAction::writeTo(OutputStream& stream) const {
    stream.write((int)(entityType));
    position.writeTo(stream);
}
bool BuildAction::operator ==(const BuildAction& other) const {
    return entityType == other.entityType && position == other.position;
}
size_t std::hash<BuildAction>::operator ()(const BuildAction& value) const {
    size_t result = 0;
    result ^= std::hash<EntityType>{}(value.entityType) + 0x9e3779b9 + (result<<6) + (result>>2);
    result ^= std::hash<Vec2Int>{}(value.position) + 0x9e3779b9 + (result<<6) + (result>>2);
    return result;
}

File: clients/CSharp/Model/BuildAction.cs
namespace Aicup2020.Model
{
    public struct BuildAction
    {
        public Model.EntityType EntityType { get; set; }
        public Model.Vec2Int Position { get; set; }
        public BuildAction(Model.EntityType entityType, Model.Vec2Int position)
        {
            this.EntityType = entityType;
            this.Position = position;
        }
        public static BuildAction ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new BuildAction();
            switch (reader.ReadInt32())
            {
            case 0:
                result.EntityType = Model.EntityType.Wall;
                break;
            case 1:
                result.EntityType = Model.EntityType.House;
                break;
            case 2:
                result.EntityType = Model.EntityType.BuilderBase;
                break;
            case 3:
                result.EntityType = Model.EntityType.BuilderUnit;
                break;
            case 4:
                result.EntityType = Model.EntityType.MeleeBase;
                break;
            case 5:
                result.EntityType = Model.EntityType.MeleeUnit;
                break;
            case 6:
                result.EntityType = Model.EntityType.RangedBase;
                break;
            case 7:
                result.EntityType = Model.EntityType.RangedUnit;
                break;
            case 8:
                result.EntityType = Model.EntityType.Resource;
                break;
            case 9:
                result.EntityType = Model.EntityType.Turret;
                break;
            default:
                throw new System.Exception("Unexpected tag value");
            }
            result.Position = Model.Vec2Int.ReadFrom(reader);
            return result;
        }
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            writer.Write((int) (EntityType));
            Position.WriteTo(writer);
        }
    }
}
